File: web/tests/browser/actor-config.py
"""Real Web/HTTP/daemon regression with an isolated home and no Actor startup.

Build web/dist and cccc first; run with --binary target/debug/cccc or a package.
The Unix IPC fixture is also run by Linux CI. No installed instance is touched.
"""
import argparse
import json
import os
from pathlib import Path
import socket
import subprocess
import tempfile
import time
import urllib.request

ROOT = Path(__file__).resolve().parents[3]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--binary", type=Path, default=ROOT / "target/debug/cccc")
    args = parser.parse_args()
    binary = args.binary.resolve()
    with tempfile.TemporaryDirectory(prefix="cc-actor-") as directory:
        temp = Path(directory)
        home = temp / "home"
        env = {**os.environ, "CCCC_HOME": str(home)}
        with socket.socket() as sock:
            sock.bind(("127.0.0.1", 0))
            port = sock.getsockname()[1]
        processes = []
        logs = []

        def launch(*argv):
            log = (temp / f"process-{len(processes)}.log").open("w+")
            logs.append(log)
            process = subprocess.Popen([str(binary), *argv], env=env, stdout=log, stderr=subprocess.STDOUT)
            processes.append(process)
            return process

        def call(op, **kwargs):
            with socket.socket(socket.AF_UNIX) as sock:
                sock.settimeout(10)
                sock.connect(str(home / "daemon/ccccd.sock"))
                sock.sendall((json.dumps({"op": op, "args": kwargs}) + "\n").encode())
                result = json.loads(sock.makefile("rb").readline())
            assert result["ok"], (op, result.get("error"))
            return result["result"]

        try:
            daemon = launch("daemon", "run")
            for _ in range(200):
                try:
                    call("ping")
                    break
                except OSError:
                    assert daemon.poll() is None, "isolated daemon exited"
                    time.sleep(0.05)
            else:
                raise AssertionError("isolated daemon did not start")
            gid = call("group_create", title="Actor configuration fixture")["group"]["group_id"]
            for key, runtime in [("source", "claude"), ("target", "codex")]:
                call("actor_profile_upsert", profile_id=key, name=f"{key.title()} profile", runtime=runtime)
            call("actor_profile_secret_update", profile_id="target", set={"PROFILE_KEY": "synthetic"})
            call("actor_add", group_id=gid, actor_id="linked", profile_id="source", enabled=False, title="Linked actor", by="user")
            command = ["codex", "-c", 'model_provider="ZAI"', "--profile", "Team One", "", "O'Reilly", r"C:\Some Folder\config"]
            call("actor_add", group_id=gid, actor_id="custom-one", runtime="codex", command=command, enabled=False, title="Custom actor", by="user", env_private={"AUDIT_OLD": "synthetic"})
            web = launch("web", "--host", "127.0.0.1", "--port", str(port))
            base = f"http://127.0.0.1:{port}"
            for _ in range(200):
                try:
                    urllib.request.urlopen(f"{base}/ui/", timeout=0.5).close()
                    break
                except OSError:
                    assert web.poll() is None, "isolated Web exited"
                    time.sleep(0.05)
            else:
                raise AssertionError("isolated Web did not start")
            metadata = temp / "fixture.json"
            metadata.write_text(json.dumps({"base": base, "gid": gid, "originalCommand": command}))
            subprocess.run(["node", str(Path(__file__).with_suffix(".mjs")), str(metadata)], cwd=ROOT / "web", env=env, timeout=180, check=True)
            assert all(not actor["enabled"] for actor in call("actor_list", group_id=gid)["actors"]), "test must never start Actors"
        except BaseException:
            for log in logs:
                log.flush()
                log.seek(0)
                print(log.read())
            raise
        finally:
            for process in reversed(processes):
                if process.poll() is None:
                    process.terminate()
            for process in reversed(processes):
                try:
                    process.wait(timeout=10)
                except subprocess.TimeoutExpired:
                    process.kill()
                    process.wait()
            for log in logs:
                log.close()


if __name__ == "__main__":
    main()

File: web/tests/browser/actor-config.mjs
import { chromium, expect } from "@playwright/test";
import fs from "node:fs";
const fixture = JSON.parse(fs.readFileSync(process.argv[2], "utf8"));
const browser = await chromium.launch({
  headless: true,
  ...(process.env.CHROME_BIN ? { executablePath: process.env.CHROME_BIN } : {}),
});
const evidence = new URL("../../test-results/actor-config/", import.meta.url);
fs.mkdirSync(evidence, { recursive: true });
const context = await browser.newContext({ viewport: { width: 1400, height: 1000 } });
await context.tracing.start({ screenshots: true, snapshots: true });
const page = await context.newPage();
let passed = false;
try {
  const api = async (path) => (await page.request.get(`${fixture.base}/api/v1/${path}`)).json();
  const actor = async (id) =>
    (await api(`groups/${fixture.gid}/actors`)).result.actors.find((a) => a.id === id);
  const dialog = page.getByRole("dialog", { name: /Edit Agent:/ });
  const openEditor = async () =>
    page.getByRole("button", { name: "Edit agent configuration", exact: true }).click();
  const save = async () => {
    await dialog.getByRole("button", { name: "Save", exact: true }).click();
    await expect(dialog).toBeHidden();
  };
  const closeActor = async () =>
    page.getByRole("button", { name: "Close expanded Actor view", exact: true }).click();
  await page.goto(`${fixture.base}/ui/`);
  await page.getByRole("button", { name: "Open terminal for Linked actor", exact: true }).click();
  await openEditor();
  await dialog.getByPlaceholder("linked", { exact: true }).fill("Linked renamed");
  await save();
  expect((await actor("linked")).profile_id).toBe("source");
  expect((await actor("linked")).title).toBe("Linked renamed");
  await openEditor();
  await dialog.getByRole("combobox", { name: "Runtime Profile", exact: true }).click();
  await page.getByRole("option", { name: /Target profile/ }).click();
  await save();
  expect((await actor("linked")).profile_id).toBe("target");
  await openEditor();
  await dialog.getByRole("button", { name: "Custom", exact: true }).click();
  await dialog.getByRole("button", { name: "Convert to Custom", exact: true }).click();
  await dialog.getByRole("combobox", { name: "Runtime", exact: true }).click();
  await page.getByRole("option", { name: "Custom", exact: true }).click();
  await dialog.locator("input.font-mono").fill("sleep 60");
  await dialog.getByRole("button", { name: "Remove PROFILE_KEY", exact: true }).click();
  let envFailures = 0;
  await page.route(`**/actors/linked/env_private`, async (route) => {
    if (route.request().method() === "POST" && envFailures++ === 0) {
      return route.fulfill({
        status: 503,
        json: {
          ok: false,
          error: { code: "fixture_failure", message: "Synthetic Actor secret failure" },
        },
      });
    }
    return route.continue();
  });
  await dialog.getByRole("button", { name: "Save", exact: true }).click();
  await expect(page.getByText(/Synthetic Actor secret failure/)).toBeVisible();
  expect((await actor("linked")).profile_id).toBe("");
  expect((await actor("linked")).runtime).toBe("custom");
  expect((await actor("linked")).command).toEqual(["sleep", "60"]);
  // Let a normal poll observe the intermediate conversion. It must not erase
  // the draft, restore the Profile mode, or clear the pending secret removal.
  await expect(dialog.getByRole("button", { name: "Save", exact: true })).toBeEnabled();
  await page.waitForTimeout(1200);
  await save();
  expect((await api(`groups/${fixture.gid}/actors/linked/env_private`)).result.keys).toEqual([]);
  expect(envFailures).toBe(2);
  await closeActor();

  await page.getByRole("button", { name: "Open terminal for Custom actor", exact: true }).click();
  await openEditor();
  await dialog.getByPlaceholder("custom-one", { exact: true }).fill("Renamed only");
  await save();
  expect((await actor("custom-one")).command).toEqual(fixture.originalCommand);
  await openEditor();
  await dialog.getByRole("button", { name: "Remove AUDIT_OLD", exact: true }).click();
  await dialog.getByRole("tab", { name: "Runtime profile tools", exact: true }).click();
  let prompts = 0;
  page.on("dialog", async (d) => {
    if (d.type() === "prompt") {
      prompts++;
      await d.accept("Saved with removal");
    } else await d.dismiss();
  });
  let profileSecretFailures = 0;
  await page.route("**/api/v1/profiles/*/env_private", async (route) => {
    if (route.request().method() === "POST" && profileSecretFailures++ === 0) {
      return route.fulfill({
        status: 503,
        json: {
          ok: false,
          error: { code: "fixture_failure", message: "Synthetic Profile secret failure" },
        },
      });
    }
    return route.continue();
  });
  await dialog.getByRole("button", { name: "Add to Runtime Profiles", exact: true }).click();
  await expect(page.getByText(/Synthetic Profile secret failure/)).toBeVisible();
  await dialog.getByRole("button", { name: "Add to Runtime Profiles", exact: true }).click();
  await expect.poll(() => profileSecretFailures).toBe(2);
  const profiles = (await api("profiles?view=accessible")).result.profiles.filter(
    (p) => p.name === "Saved with removal",
  );
  expect(profiles).toHaveLength(1);
  expect(prompts).toBe(1);
  expect(profiles[0].command).toEqual(fixture.originalCommand);
  await expect
    .poll(
      async () => (await api(`profiles/${profiles[0].id}/env_private?scope=global`)).result.keys,
    )
    .toEqual([]);
  await dialog.getByRole("button", { name: "Cancel", exact: true }).click();
  await closeActor();
  // Closing an inbox does not cancel its HTTP response. A later opening owns
  // the content even when the same Actor is opened again.
  await page.getByRole("button", { name: "Open terminal for Linked renamed", exact: true }).click();
  let releaseOldInbox;
  let inboxRequests = 0;
  await page.route("**/inbox/linked?*", async (route) => {
    const index = ++inboxRequests;
    if (index === 1)
      await new Promise((resolve) => {
        releaseOldInbox = resolve;
      });
    await route.fulfill({
      json: {
        ok: true,
        result: {
          messages: [
            {
              id: `inbox-${index}`,
              kind: "chat.message",
              by: "user",
              ts: "2026-09-18T00:00:00Z",
              data: { text: index === 1 ? "Previous inbox snapshot" : "Current inbox snapshot" },
            },
          ],
        },
      },
    });
  });
  await page.getByRole("button", { name: /^Open inbox/ }).click();
  const inbox = page.getByRole("dialog", { name: /^Mail · linked/ });
  await expect(inbox).toBeVisible();
  await expect.poll(() => !!releaseOldInbox).toBe(true);
  await inbox.getByRole("button", { name: "Close", exact: true }).click();
  await page.getByRole("button", { name: /^Open inbox/ }).click();
  await expect(inbox).toContainText("Current inbox snapshot");
  const oldResponse = page.waitForResponse((response) => response.url().includes("/inbox/linked?"));
  releaseOldInbox();
  await oldResponse;
  // Observe beyond delivery of the response and React's update.
  await page.waitForTimeout(150);
  await expect(inbox).toContainText("Current inbox snapshot");
  await expect(inbox).not.toContainText("Previous inbox snapshot");
  await inbox.getByRole("button", { name: "Close", exact: true }).click();
  await page.unroute("**/inbox/linked?*");

  let releaseDelete;
  await page.route("**/actors/linked?*", async (route) => {
    if (route.request().method() === "DELETE") {
      await new Promise((resolve) => {
        releaseDelete = resolve;
      });
    }
    await route.continue();
  });
  page.removeAllListeners("dialog");
  page.on("dialog", (d) => (d.type() === "confirm" ? d.accept() : d.dismiss()));
  await page.getByRole("button", { name: "Remove agent", exact: true }).click();
  await expect.poll(() => !!releaseDelete).toBe(true);
  await closeActor();
  await page.getByRole("button", { name: "Open terminal for Renamed only", exact: true }).click();
  const deleted = page.waitForResponse(
    (response) =>
      response.request().method() === "DELETE" &&
      new URL(response.url()).pathname.endsWith("/actors/linked"),
  );
  releaseDelete();
  await deleted;
  await page.evaluate(
    () => new Promise((done) => requestAnimationFrame(() => requestAnimationFrame(done))),
  );
  await expect.poll(() => actor("linked")).toBeUndefined();
  await expect(
    page.getByRole("button", { name: "Close expanded Actor view", exact: true }),
  ).toBeVisible();
  await expect(
    page.getByRole("button", { name: "Edit agent configuration", exact: true }),
  ).toBeVisible();
  await closeActor();
  await page.getByRole("button", { name: "Settings and more", exact: true }).click();
  await page.getByRole("button", { name: "This instance", exact: false }).click();
  await page.getByRole("button", { name: "Actor Profiles", exact: true }).click();
  const card = page.getByText("Source profile", { exact: true }).locator("xpath=../../..");
  await card.getByRole("button", { name: "Edit", exact: true }).click();
  const editor = page
    .getByRole("dialog")
    .filter({ has: page.getByText("Edit Actor Profile", { exact: true }) })
    .last();
  await editor.locator("textarea").first().fill("AUDIT_NEW=synthetic");
  let failures = 0;
  await page.route("**/api/v1/profiles/source/env_private", async (route) => {
    if (route.request().method() === "POST" && failures++ === 0) {
      return route.fulfill({
        status: 503,
        json: {
          ok: false,
          error: { code: "fixture_failure", message: "Synthetic settings secret failure" },
        },
      });
    }
    return route.continue();
  });
  await editor.getByRole("button", { name: "Save", exact: true }).click();
  await expect(editor.getByRole("alert")).toContainText("Synthetic settings secret failure");
  expect((await api("profiles/source")).result.profile.revision).toBe(2);
  await editor.getByRole("button", { name: "Save", exact: true }).click();
  await expect(editor).toBeHidden();
  expect(failures).toBe(2);
  expect((await api("profiles/source/env_private?scope=global")).result.keys).toEqual([
    "AUDIT_NEW",
  ]);
  passed = true;
  process.stdout.write(
    "Actor configuration real-port workflows passed: linked rename/switch/convert, draft retention, exact argv, effective secrets, partial-save retry without duplicates or stale revisions, inbox response ownership and delayed deletion preserving navigation.\n",
  );
} finally {
  if (!passed)
    await page
      .screenshot({ path: new URL("failure.png", evidence).pathname, fullPage: true })
      .catch(() => {});
  await context.tracing.stop(passed ? {} : { path: new URL("trace.zip", evidence).pathname });
  await browser.close();
}
